---
title: "2.2 Assembly"
---

**Teaching: 40 min** || **Exercises: 20 min**


## Overview

:::::{.callout}

:::{.callout-important icon=false}
### Questions:
- Insert Questions Here
- Insert Questions Here
- Insert Questions Here
:::

:::{.callout-important icon=false}
### Learning Objectives:
- Insert Learning Objectives Here
- Insert Learning Objectives Here
- Insert Learning Objectives Here
:::

:::{.callout-tip}
### Key Points:
- Insert Key Points Here
- Insert Key Points Here
- Insert Key Points Here
  - You can Introduce an indentation
  - You can Introduce an indentation
    - You can Introduce an indentation
:::
:::::


<!--
Flye

Nanopore assemblers are still in development, but Flye is a de facto standard. Installation:

# Note: not on 22.04 yet, works from 23.04
sudo apt install flye

Running Flye is very simple, and according to its [documentation](https://github.com/fenderglass/Flye) needs no prior cleaning or trimming of the reads:

# Using all CPUs on the machine (nproc)
flye -t $(nproc) --out-dir $OUT_DIR --nano-raw $INPUT_FQ

If the reads were corrected, use --nano-corr instead of --nano-raw:

flye -t $(nproc) --out-dir $OUT_DIR --nano-corr $INPUT_FQ

When basecalling was done with the Guppy SUP model and/or read quality is Q20+:

flye -t $(nproc) --out-dir $OUT_DIR --nano-hq $INPUT_FQ



Medaka

Medaka is an assembly polisher and variant caller made by ONT. It is the recommended polisher for Flye assemblies. The Medaka [documentation](https://github.com/nanoporetech/medaka) mentions that is has specifically been trained on Flye output (it is an ML-based tool).

Installation ideally from BioConda:

mamba create -n medaka -c conda-forge -c bioconda medaka

If this does not work, or the version you obtain is out of date, building from source in a Conda environment (see the Chapter on “Installation Do’s and Donts” below) is your best option.

To run medaka (using all CPU’s):

conda activate medaka
medaka_consensus -i INPUT_FQ -d ASSEMBLY_FA -o OUTDIR -m MEDAKA_MODEL -t $(nproc)

Where MEDAKA_MODEL should be chosen as closely as possible to the model used for guppy_basecaller. medaka tools list_models lists the available models.

If Medaka runs out of memory, add option -b 80 to the command-line. If it still runs out of memory, reduce the 80 further until it doesn’t.

The Medaka docs mention that”for users with RTX series GPUs it may be required to additionally set:

# Execute this AFTER conda activate medaka
export TF_FORCE_GPU_ALLOW_GROWTH=true

Medaka also has a CPU version, which should work in all cases but is much slower.



## Assembly QC

Quast 




NB. Once FastQC and Quast have both run, use MultiQC to integrate their output into one HTML report.

-->



## 1.1 Insert First Topic Here

Contents of topic goes here

You can insert a video from YouTube using the below format

<p align="center"><iframe width="560" height="315" src="https://www.youtube.com/embed/mI0Fo9kaWqo" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe></p>

### sub-heading

Insert your contents here.


### another sub-heading

Insert your contents here.

Use the below format to introduce note callouts

:::{.callout-note}
**heading**

Insert your contents here.
:::


## 1.2 Insert Second Topic Here


### sub-heading

Insert your contents here.

The below format allows you to display codes in their natural way:

```
@SEQ_ID                   <-- SEQUENCE NAME
AGCGTGTACTGTGCATGTCGATG   <-- SEQUENCE
+                         <-- SEPARATOR
%%).1***-+*''))**55CCFF   <-- QUALITY SCORES
```

Below is another format for callouts

:::{.callout}
### Insert callout heading here

Insert your contents here.

The format below helps display your bash codes just like they will appear in the command line
```bash
cd ~/Desktop/course_files/01_intro/
```

:::


Below is how to point to an image/figure in your ***fig*** directory.

![FastQC quality report](../fig/fastqc_quality.png)

We will explore more on *FASTQC* later.



### MORE FORMARTING INSTRUCTIONS

## Callouts

:::{.callout}
This is a callout without specification
:::

:::{.callout}
### *output:*
This is a callout without specification
:::

:::{.callout-note}
Note that there are five types of callouts, including: 
`note`, `tip`, `warning`, `caution`, and `important`.
:::

:::{.callout-tip}
This is a callout `tip`
:::

:::{.callout-warning}
This is a callout `warning`
:::

:::{.callout-caution}
This is a callout `caution`
:::

:::{.callout-important}
This is a callout `important`.
:::

:::{.callout-important collapse="true"}
This is a collapsible callout.
:::


## Source codes

python

```python
1 + 1
```

bash
```bash
1 + 1
```

## General syntax for displaying codes and output in a callout
:::{.callout}
```bash
$ ls -F
```
```
insert output here
```
:::


## EXERCISES & HIDDEN SOLUTIONS
:::::{.callout-important icon=false}
## ***Exercise:*** Insert title of exercise
Content of exercise goes here

:::{.callout collapse="true"}
## ***Solution:***
Content of solution goes here
:::
:::::


## Hiding notes
<details><summary>Hint</summary>Insert notes here.</details>


## Referencing figures/images
![Caption](path_to_image.png)



## 1.3 Credit
Information on this page has been adapted and modified from the following source:

- 